src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+
+	// Set lo speed for initialization
+	everdrive_sd_config &= ~ED64_SD_CFG_SPEED;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+
+	// Put in idle
+	everdrive_sd_execute_command(ED64_SD_CMD0, 0);
+
+	// IF cond
+	result = everdrive_sd_execute_command(ED64_SD_CMD8, 0x1AA);
+
+	if(result & ED64_SD_TIMEOUT) return RES_ERROR;
+
+	if (everdrive_sd_resp_buff[3] != 1) {
+		debugf("SD card - voltage mismatch\n");
+		return RES_ERROR;
+	}
+
+	if(everdrive_sd_resp_buff[4] != 0xAA) {
+		debugf("SD card did not echo\n");
+		return RES_ERROR;
+	};
+
+	int num_retries = ED64_SD_ACMD41_TOUT_MS / ED64_SD_ACMD41_WAIT_MS;
+	while(1) {
+		if(!num_retries--) {
+			debugf("SD card did not respond\n");
+			return RES_ERROR;
+			break;
+		}
+
+		// Query with HCS and 3.2-3.4V
+		result = everdrive_sd_send_app_command(ED64_SD_CMD41, 0, 0x40300000);
+
+		// Check ready bit on OCR
+		if ((everdrive_sd_resp_buff[1] & 0x80) != 0) break;
+
+		wait_ms(ED64_SD_ACMD41_WAIT_MS);
+	};
+
+	// Check CCS and set HC mode
+	everdrive_sd_active_mode |= everdrive_sd_resp_buff[1] & 0x40;
+
+	result = everdrive_sd_execute_command(ED64_SD_CMD2, 0);
+	if (result & ED64_SD_TIMEOUT) return RES_ERROR;
+
+	result = everdrive_sd_execute_command(ED64_SD_CMD3, 0);
+	if (result & ED64_SD_TIMEOUT) return RES_ERROR;
+
+	sd_rca = 	(everdrive_sd_resp_buff[1] << 24) |
+				(everdrive_sd_resp_buff[2] << 16) |
+				(everdrive_sd_resp_buff[3] << 8) |
+				everdrive_sd_resp_buff[4];
+
+	result = everdrive_sd_execute_command(ED64_SD_CMD7, sd_rca);
+	if (result & ED64_SD_TIMEOUT) return RES_ERROR;
+
+	// Set bus width to 4
+	result = everdrive_sd_send_app_command(ED64_SD_CMD6, sd_rca, 0x2);
+	if (result & ED64_SD_TIMEOUT) {
+		debugf("ACMD6 err\n");
+		return RES_ERROR;
+	};
+
+	// Set hi speed
+	everdrive_sd_config |= ED64_SD_CFG_SPEED;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+
+	return 0;
+}
+
+static DRESULT fat_disk_read_everdrive(BYTE* buff, LBA_t sector, UINT count)
+{
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	uint8_t crc[8];
+
+	// Overclock the PI
+	uint32_t old_pw = PI_regs->dom1_pulse_width;
+	PI_regs->dom1_pulse_width = 0x09;
+
+	if(everdrive_sd_change_mode(ED64_SD_MODE_BLOCK_READ, sector)) {
+		PI_regs->dom1_pulse_width = old_pw;
+		return RES_ERROR;
+	};
+
+	for (int i=0;i<count;i++)
+	{
+		uint16_t timeout = -1;
+		uint8_t result;
+		// Each 1 bit everdrive_sd_read_data shifts 4 bit of data from the four
+		// data lanes. To find the start marker, wait for all lanes to go low to
+		// start the transfer.
+		set_everdrive_sd_bitlen(1);
+		do {
+			if (!timeout--) {
+				PI_regs->dom1_pulse_width = old_pw;
+				debugf("Data token timeout\n");
+				return RES_ERROR;
+			}
+			result = everdrive_sd_read_data();
+		} while (result != 0xf0);
```

Comment:
There are many of these loops in the code. Maybe you could factor them out into a function?

Reply:
That's good advice but there are slight nuances that makes it a little convoluted. So I'll leave them as is for now.